### PR TITLE
Refactor to include Slot struct

### LINKED ISSUE #152: Refactor to include a Slot struct
While working on #150 , I realized there's a need for some refactoring. To start with, we can benefit from a `Slot` struct that will encapsulate some functionality relating to working with time slots.  Currently we're working directly with tuples of `(NaiveDateTime,NaiveDateTime)` and this is becoming trickier as we add more complex features to the scheduler.

The `Slot` struct will at its core represent a contiguous block of time, defined by a start NaiveDateTime, an end NaiveDateTime plus useful methods for working with time slots.

### PR DESCRIPTION
**Pull request by egithinji**
This closes #152 . More refactoring will follow this but will open separate issues for those other changes.

### CODE REVIEW

File: src/slot.rs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    pub start: NaiveDateTime,
+    pub end: NaiveDateTime,
+}
+
+impl Sub for Slot {
+    type Output = Vec<Slot>;
+
+    fn sub(self, other: Self) -> Self::Output {
+        let mut result = Vec::new();
+        if (other.start < self.start) && (other.end <= self.start) {
```
Comment (tijlleenders) [inline src/slot.rs:16]: I would expect six different cases:
- entirely before
- entirely after
- overlapping begin
- overlapping end
- overlapping begin + end (equal or bigger)
- overlapping within (smaller)

Mind the equals for the slot [begin:end[:
- begin is inclusive
- end is exclusive

**Review by tijlleenders — CHANGES_REQUESTED** [review]

File: src/slot.rs (hunk 1 above)
Comment (tijlleenders) [reply src/slot.rs:16]: Subtraction - I count seven if blocks.
Didn't check details.

Comment (egithinji) [reply src/slot.rs:16]: I've just double-checked and I believe all cases are handled. You can verify from the comments in the `if` blocks or from the unit tests from line 377.

Comment (tijlleenders) [reply src/slot.rs:16]: If we can simplify to 6 if blocks instead of 8 I think we should.

**Comment by tijlleenders** [comment]
Something to think about after this PR: a Task has a Vec<Slot>. This vec can hold none, one, or many Slots. Slot arithmetic should probably take a Vec and output a Vec - or it will lead to inconsistencies which make chaining impossible.

**Review by tijlleenders — APPROVED** [review]

**Pull request merged**